Implements:
* Automatic parenthesis matching.

Applying the patch:
1. Move/copy the patchfile to

   p9p/../src/cmd/

2. cd to p9p/../src/cmd/acme/
3. run

   patch -Np1 < ../patchfile 2>&1 | tee log

Building acme:
1. cd to p9p/../src/cmd/amce/
2. run

   ../../../bin/9 mk install 2>&1 | tee log

3. (Optionial) when the build was succesful or you want to
   start over from scratch:

   ../../../bin/9 mk clean

--- acme.orig/text.c	2013-10-23 11:00:07.000000000 +0200
+++ acme.paren/text.c	2016-02-23 11:37:15.000000000 +0100
@@ -653,6 +653,10 @@
 	return rp;
 }
 
+static int findmatch = 0;
+static uint oldpos = 0;
+static uint matchp0, matchp1;
+
 void
 texttype(Text *t, Rune r)
 {
@@ -666,6 +670,34 @@
 		return;
 	if(t->what == Tag)
 		t->w->tagsafe = FALSE;
+   
+	/* Clear (possible) selection if we did parenmatching. */
+	if (findmatch == 1) {
+		typecommit(t);
+		/*
+		 * Bug: If the user selects the paren-match selection itself--
+		 *    selects the selection-- we _will_ return to oldpos,
+		 *    since this `new' selection is the same as the one that
+		 *    was a result of paren-matching.  This implementation
+		 *    cannot tell the difference. No select-and-type-to-overwrite
+		 *    is possible in this case. Probably rare, but a bug
+		 *    nonetheless.
+		 *
+		 *   :/findmatch\(#/ 
+		 */
+		if(t->q1 < t->file->b.nc && t->fr.p0 != t->fr.p1) {
+			if(t->fr.p0 == matchp0 && t->fr.p1 == matchp1){
+				textsetselect(t, t->q1, t->q1);
+				textshow(t, oldpos, oldpos, TRUE);
+			}
+		}else{
+			textsetselect(t, t->q1, t->q1);
+			textshow(t, t->q1, t->q1, TRUE);
+		}
+		/* set to current */
+		oldpos = t->q1;
+		findmatch = 0;
+	}
 
 	nr = 1;
 	rp = &r;
@@ -879,6 +911,16 @@
 			}
 		}
 		break; /* fall through to normal code */
+	
+	case ')':  /* fallthrough */
+	case '}':  /* fallthrough */
+	case ']':  /* fallthrough */
+	case '>':  /* fallthrough */
+	case '\'': /* fallthrough */
+	case '"':  /* fallthrough */
+	case '`': 
+		findmatch = 1;
+		break;
 	}
 	/* otherwise ordinary character; just insert, typically in caches of all texts */
 	for(i=0; i<t->file->ntext; i++){
@@ -915,6 +957,37 @@
 	if(r=='\n' && t->w!=nil)
 		wincommit(t->w, t);
 	t->iq1 = t->q0;
+
+	if(findmatch == 1){   /* :/findmatch\(#/ */
+		/*
+		 * Put cursor on the left ("inside") of paren-char.
+		 */
+		q0 = t->q0-1;
+		q1 = t->q1-1;
+		oldpos = t->q1;
+		typecommit(t);
+		textdoubleclick(t, &q0, &q1);
+ 		/* found a matching character */
+		if(q0 != q1){
+			/*
+			 * If the matching char falls outside the visual
+			 * text (is `invisible') do nothing.
+			 */
+			if(q0 < t->org){
+				findmatch=0;
+			}else{
+				textsetselect(t, q0-1, q0);
+				/* 
+				 * Set matchN to fr.p0/p1 not q0-1/q0, since
+				 * t->q0/q1 may be off. See textsetselect().
+				 */ 
+				matchp0=t->fr.p0;
+				matchp1=t->fr.p1;
+			}
+		}
+	}else{ /* redundant, but just to be sure */
+		findmatch = 0;
+	}
 }
 
 void
